adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,712 @@
+        // LogInfo("result is 0x%x\n", result);
+
+        if (result == NULL)
+        {
+            LogError("Failed allocating TLSIO instance.");
+        }
+        else
+        {
+            memset(result, 0, sizeof(TLS_IO_INSTANCE));
+            mallocAndStrcpy_s(&result->hostname, tls_io_config->hostname);
+            result->port = tls_io_config->port;
+            result->ssl_context = NULL;
+            result->ssl = NULL;
+            result->certificate = NULL;
+
+            result->on_bytes_received = NULL;
+            result->on_bytes_received_context = NULL;
+
+            result->on_io_open_complete = NULL;
+            result->on_io_open_complete_context = NULL;
+
+            result->on_io_close_complete = NULL;
+            result->on_io_close_complete_context = NULL;
+
+            result->on_io_error = NULL;
+            result->on_io_error_context = NULL;
+
+            result->tlsio_state = TLSIO_STATE_NOT_OPEN;
+
+            result->x509certificate = NULL;
+            result->x509privatekey = NULL;
+        }
+    }
+
+    return (CONCRETE_IO_HANDLE)result;
+}
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_010: [ The tlsio_openssl_destroy succeed ]*/
+void tlsio_openssl_destroy(CONCRETE_IO_HANDLE tls_io)
+{
+    /* Codes_SRS_TLSIO_SSL_ESP8266_99_009: [ The tlsio_openssl_destroy NULL parameter. make sure there is no crash ]*/
+    if (tls_io == NULL)
+    {
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+        if (tls_io_instance->certificate != NULL)
+        {
+            free(tls_io_instance->certificate);
+            tls_io_instance->certificate = NULL;
+        }
+        if (tls_io_instance->hostname != NULL)
+        {
+            free(tls_io_instance->hostname);
+        }
+        if (tls_io_instance->x509certificate != NULL)
+        {
+            free((void*)tls_io_instance->x509certificate);
+        }
+        if (tls_io_instance->x509privatekey != NULL)
+        {
+            free((void*)tls_io_instance->x509privatekey);
+        }
+        free(tls_io);
+    }
+}
+
+
+/* Codes_SRS_TLSIO_SSL_ESP8266_99_008: [ The tlsio_openssl_open succeed ]*/
+int tlsio_openssl_open(CONCRETE_IO_HANDLE tls_io, ON_IO_OPEN_COMPLETE on_io_open_complete, void* on_io_open_complete_context, ON_BYTES_RECEIVED on_bytes_received, void* on_bytes_received_context, ON_IO_ERROR on_io_error, void* on_io_error_context)
+{
+    int result;
+
+    if (tls_io == NULL)
+    {
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_006: [ The tlsio_openssl_open failed because tls_io is NULL. ]*/
+        result = __LINE__;
+        LogError("NULL tls_io.");
+    }
+    else
+    {
+
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        /* Codes_SRS_TLSIO_SSL_ESP8266_99_007: [ The tlsio_openssl_open invalid state. ]*/
+        if (tls_io_instance->tlsio_state != TLSIO_STATE_NOT_OPEN)
+        {
+            result = __LINE__;
```

Comment:
If callbacks was provided, you probably need to call on_io_error and on_io_open_complete with IO_OPEN_ERROR.